### ISSUE TITLE
JAX JIT Compilation Issue with jnp.std and Configuration Setting

### ISSUE THREAD

**Issue by TomasAndersonFang** [google/jax#22308]
### Description

When I ran the following code, I got the following error. But when I commented `config.update('jax_disable_jit', True)
`, everything is ok. Intuitionally, commenting or not commenting should have no influence on the code. I also tried other APIs like mean or add to replace std, but they didn't trigger the same problem.

```python
from jax import config
config.update("jax_debug_nans", True)
config.update('jax_disable_jit', True)

import jax.numpy as jnp
import jax
import pytest

x = jnp.arange(0, 0.5, step=0.1)

@jax.jit
def f(x):
    z = jnp.std(x)
    return z

def test_f():
    f(x)

test_f()

```
Error:
```
Traceback (most recent call last):
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/pjit.py", line 1568, in _pjit_call_impl_python
    return compiled.unsafe_call(*args), compiled
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/profiler.py", line 335, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py", line 1258, in __call__
    dispatch.check_special(self.name, arrays)
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/dispatch.py", line 315, in check_special
    _check_special(name, buf.dtype, buf)
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/dispatch.py", line 320, in _check_special
    raise FloatingPointError(f"invalid value (nan) encountered in {name}")
FloatingPointError: invalid value (nan) encountered in jit(convert_element_type)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tomasandersonfang/Desktop/reserach/dnnbugs/dataset/test_script.py", line 19, in <module>
    test_f()
  File "/Users/tomasandersonfang/Desktop/reserach/dnnbugs/dataset/test_script.py", line 17, in test_f
    f(x)
  File "/Users/tomasandersonfang/Desktop/reserach/dnnbugs/dataset/test_script.py", line 13, in f
    z = jnp.std(x)
        ^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/reductions.py", line 506, in std
    return _std(a, _ensure_optional_axes(axis), dtype, out, correction, keepdims,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/reductions.py", line 519, in _std
    return lax.sqrt(var(a, axis=axis, dtype=dtype, correction=correction, keepdims=keepdims, where=where))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/reductions.py", line 441, in var
    return _var(a, _ensure_optional_axes(axis), dtype, out, correction, keepdims,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/reductions.py", line 474, in _var
    return _where(normalizer > 0, lax.div(result, normalizer).astype(dtype), np.nan)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/util.py", line 449, in _where
    x, y = promote_dtypes(x, y)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/util.py", line 278, in promote_dtypes
    return [lax._convert_element_type(x, to_dtype, weak_type) for x in args]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/numpy/util.py", line 278, in <listcomp>
    return [lax._convert_element_type(x, to_dtype, weak_type) for x in args]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/lax/lax.py", line 559, in _convert_element_type
    return convert_element_type_p.bind(operand, new_dtype=new_dtype,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/core.py", line 416, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/core.py", line 420, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/core.py", line 921, in process_primitive
    return primitive.impl(*tracers, **params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tomasandersonfang/opt/anaconda3/envs/dnn-test/lib/python3.11/site-packages/jax/_src/dispatch.py", line 87, in apply_primitive
    outs = fun(*args)
           ^^^^^^^^^^
FloatingPointError: invalid value (nan) encountered in jit(convert_element_type). Because jax_config.debug_nans.value and/or config.jax_debug_infs is set, the de-optimized function (i.e., the function as if the `jit` decorator were removed) was called in an attempt to get a more precise error message. However, the de-optimized function did not produce invalid values during its execution. This behavior can result from `jit` optimizations causing the invalid value to be produced. It may also arise from having nan/inf constants as outputs, like `jax.jit(lambda ...: jax.numpy.nan)(...)`.

It may be possible to avoid the invalid value by removing the `jit` decorator, at the cost of losing optimizations.

If you see this error, consider opening a bug report at https://github.com/google/jax.
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.30
jaxlib: 0.4.30
numpy:  1.26.4
python: 3.11.9 (main, Apr 19 2024, 11:43:47) [Clang 14.0.6 ]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='MacBook-Pro.local', release='23.4.0', version='Darwin Kernel Version 23.4.0: Fri Mar 15 00:19:22 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T8112', machine='arm64')
```

**Comment by superbobry** [comment]
Thanks! Minimal repro:

```
import jax
import jax.numpy as jnp

jax.config.update('jax_debug_nans', True)
jax.config.update('jax_disable_jit', True)

x = jnp.arange(0, 0.5, step=0.1)
jax.jit(jnp.var)(x)
```

@yashk2810 can you take a look, please?

**Comment by jakevdp** [comment]
I got this one – fix in #22312
